[RNMobile] Add "Set as Featured Image" Button to Image Block (iOS Only)

SiobhyB opened the pull request:
Partial fix for: wordpress-mobile/gutenberg-mobile#1011

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3450
`WordPress-iOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/16426

The main PR for the branch this individual PR will be merged into, `add/featured-functionality-to-image-block-ios`, can be found here: #31345

## Description

Building on the work done in https://github.com/WordPress/gutenberg/pull/31347, this PR will add a **Set as Featured** button to the image block's settings, with the purpose being to make it simpler for users to set a featured image within the post's editor. Users will also be able to **Remove as Featured** directly from the block's settings. 

## How has this been tested?

### Test Case 1: General Flow

<details>
<summary>
To start with, the general flow for setting a featured image via the image block on a post without an existing image is as follows.</summary>

* In the app, navigate to _Home_ → _Blog Posts_ to create a new blog post, one which doesn't already have a featured image set.
* Add a new image block, and upload an image to it.
* Tap the image block's cog/gear icon to access its settings.
* Observe the new **Set as Featured** button at the bottom of the settings and tap it.
* Verify that you see a confirmation message about the newly set featured image and that the block now has a **Featured** banner overlaying it.</details>

### Test Case 2: "Replace Existing" Flow

<details>
<summary>
It's also possible for users to replace an existing featured image via the following flow.</summary>

* Add a new image block to the post and upload a second image to it.
* Again, tap the image block's cog/gear icon to access its settings.
* Tap the **Set as Featured** button at the bottom of the settings and tap it.
* A dialog should appear to inform you that the post already has a featured image, with an option to **Cancel** or confirm your wish to **Replace**. 
* Tap **Cancel** to confirm it works as expected, to dismiss the action.
* Go through steps two through four again and this time tap **Replace**.
* Verify that you see a confirmation message about the newly set featured image and that the block now has a **Featured** banner overlaying it. The banner should no longer overlay the previous featured image.</details>

### Test Case 3: "Editor Loads" Flow

<details>
<summary>Next, we should verify that a **Remove as Featured** option appears when a post with a featured image is first loaded. To test this, follow these steps.</summary>

* Verify that a featured image has been set on the post and tap the back button.
* Return to the post's editor by tapping on it from the post list screen.
* Verify that the post's featured image has a **Featured** banner overlaying it. Tap the cog/gear icon to then confirm the option to **Remove as Featured** appears.</details>

### Test Case 4: "Post Settings" Flow

<details>
<summary>The image block should be updated to reflect whether its image is featured or not even when the featured image is set via the general Post Settings flow.</summary>

* Tap the three icons to the upper right of the post editor to access **Post Settings**.
* Tap existing featured image and then **Remove**.
* Tap the back button to confirm that the **Featured** banner has been removed from the featured image, as expected. Tap the cog/gear icon to then confirm the option to **Set as Featured** appears.
* Return to the **Post Settings** screen and set a new featured image.
* Tap the back button. If the new featured image hasn't already been added to the editor via an image block, add it.  
* Confirm that the **Featured** banner displays overlays the correct image and that  **Remove as Featured** appears in its settings.</details>

### Test Case 5: "Replace in Image Block" Flow

<details>
<summary>Finally, it's possible to replace an image directly within the image block's settings. We need to verify that the featured settings update as expected depending on the image that's chosen for an image block.</summary>

* Select any image block that is flagged with the **Featured** banner.
* Tap the edit icon to the upper right and then **Replace**. 
* Replace with any image that _isn't_ the post's featured image.
* Verify that the **Featured** banner is removed from the block when the new image uploads and that the **Set as Featured** button appears in the block's settings.
* Go through the steps to replace the block's image again, this time pick the post's featured image.
* Verify that the **Featured** banner appears again and that the **Remove as Featured** button appears in the block's settings.</details>

## Screenshots <!-- if applicable -->

| Set as Featured Image  | Remove as Featured Image |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/116890994-1ff7cb80-ac26-11eb-9ff9-52ff0baea855.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/116890858-f8a0fe80-ac25-11eb-96d3-d220a060e1cd.png" width="100%"> |

| Replace Dialog  | Confirmation Dialog |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/116890946-12dadc80-ac26-11eb-9b41-bd043400c39d.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/116890771-e3c46b00-ac25-11eb-9407-cf84e5f700e0.png" width="100%"> |

## Types of changes

This PR introduces a new feature (non-breaking change which adds functionality). The feature it introduces is a **Set as Featured** button in the image block's setting, which enables a user to either set an image as featured or remove it as featured. A high-level overview of the code changes involved is as follows:

* This PR relies on the `getSetFeaturedButton` component deployed in the Android version of this PR (https://github.com/WordPress/gutenberg/pull/28854) to display a `Set as Featured` or a `Remove as Featured` button, depending on whether `isFeaturedImage` is true or not. Further details around the logic behind `isFeaturedImage` can also be found in that PR. 
* `setFeaturedImage` will be called when the button is tapped, this function utilizes another function named `gutenbergDidRequestToSetFeaturedImage` to [send an image's media ID over the bridge](https://github.com/WordPress/gutenberg/pull/31415/files#diff-6dfc6ea3ebbdb9e3cfd6f9501f7837147ed5352071ffde792b4704ed387c5dd5R133) when a request is made to either set a new featured image (in which case the media ID is that of the selected image) or to remove an existing featured image (in which case the media ID is zero).
* When the featured image is changed, the native side will update Gutenberg via the `featuredImageIdNativeUpdated` listener, keeping the native apps as our main source of truth when it comes to confirming the change to the post's featured image ID.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Set up "featuredImageId" as "initialProps" and send gutenbergFeaturedImageId() over the bridge

gutenbergFeaturedImageId() will be used to gather the initial ID of a post's featured image when the editor first loads in iOS.

SiobhyB committed:
Set up "featuredImageIdNativeUpdated" to fetch details of an updated featured image over the bridge

SiobhyB committed:
Removed isAndroid conditional surrounding "featuredImageIdNativeUpdated" listener

SiobhyB committed:
Update style.css to ensure badge is rounded on iOS

"border-radius" doesn't function as intended on iOS devices without use of "overflow: hidden".

SiobhyB committed:
Add gutenbergFeaturedImageId() to GutenbergDemo/GutenbergViewController.swift

This is to align the code in the demo file with the code in GutenbergBridgeDataSource, so that it runs as expected.

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/featured-badge-ios

SiobhyB committed:
Remove androidOnly conditional surrounding "set as featured" button

SiobhyB committed:
Add "setFeaturedImage" method to bridge for iOS

SiobhyB committed:
Update gutenbergDidRequestToSetFeaturedImage() in GutenbergDemo/GutenbergViewController

This update ensures the function conforms to the pattern set out for it in GutenbergBridgeDelegate

SiobhyB committed:
Update WPAndroidGlueCode with setFeaturedImage() bridge functions

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

guarani commented:
Testing https://github.com/wordpress-mobile/WordPress-iOS/pull/16426/commits/1a8758a9d12e69932288e47c2e19f0a1df978cd2 on iPhone 11 Pro Max simulator.

### Test Case 1: General Flow

- The "Set as Feature Image" option is off-screen for me on iPhone 11 Pro Max simulator when using an image that **had a link set**, so its discoverability seems low in certain situations (which reminded me of [this post](https://tyler.io/perfectly-cropped/) 😄).
- The small edit/replace button that was previously shown _inside_ an image canvas is now shown outside. ~~I have to check `trunk` to see if this is specific to this branch or is a change or regression on `trunk`.~~ It looks like this bug isn't present on `trunk`:

  | This branch | `trunk` |
  | - | - |
  | <img width="306" alt="Screen Shot 2021-06-01 at 15 36 21" src="https://user-images.githubusercontent.com/1898325/120380714-82a8b980-c2ef-11eb-9aa1-3c818cf9851a.png"> | <img width="307" alt="Screen Shot 2021-06-01 at 18 03 43" src="https://user-images.githubusercontent.com/1898325/120395914-c86f7d00-c303-11eb-860d-283c615c3e52.png"> |


- When the message "Set/Removed as Feature Image" is shown for a few seconds, just before it's removed, it seems to flash momentarily. I'm not sure if this is a symptom of something being called twice for example, so I just want to flag it for you (at 0:05s in the video below):

  <details>
    <summary>See video</summary>

    https://user-images.githubusercontent.com/1898325/120381613-addfd880-c2f0-11eb-8aec-a78012b36c7a.mov
  </details>

- I noticed that featured images didn't load in post previews on _private sites_, but it has nothing to do with this PR and I filed an issue here: https://github.com/wordpress-mobile/WordPress-iOS/issues/16607

### Test Case 2: "Replace Existing" Flow

- The bottom (default) button in the action sheet that's shown when replacing a featured image should be "Cancel" At the moment, both options shown are "Replace"

  <details>
    <summary>See screenshot</summary>
    <img width="301" alt="Screen Shot 2021-06-01 at 16 34 24" src="https://user-images.githubusercontent.com/1898325/120387549-0e264880-c2f8-11eb-899e-7bb1e59c1f13.png">
  </details>
- Looking at the screenshot above, I'm not sure if the action sheet's title should be a question (i.e. "Replace?"). Looking at other parts of the app, I notice the edit/replace button's action sheet doesn't use a question in its title. An alternative could be "Featured image already set" (or "Featured image options"). I'm fine leaving it as-is, just wanted to share the thought.
- When a post already has a featured image set via Post Settings and the user goes to set a different image via the Image block and is presented with the replace action sheet, I think some users might be confused because they don't see any other image in the post marked as Featured. This could happen for users who aren't already familiar with Post Settings (e.g. they haven't used featured images in Post Settings before, or they've only ever set featured images on the web, etc). One option here would be to make the replace action sheet title explain where the image is already set. This is more of an enhancement (not a blocker for this PR), but to give an example, the title could be:
  - _"Featured image already set in Post Settings"_ if image was set via Post Settings, or
  - _"Featured image already set"_ if image was set via an Image block
- "Feature Image" in the "Set/Removed as Feature Image" messages maybe should be lowercase i.e. "Set/Removed as feature image" (it's up to you, I'm open to what you think is best here).

### Test Case 3: "Editor Loads" Flow

No new feedback.

### Test Case 4: "Post Settings" Flow

No new feedback.

### Test Case 5: "Replace in Image Block" Flow

- Here things worked as described in the test case, but my expectation was that replacing a featured image would update the post's featured image to the new image. In this scenario I can think of three options:
  A. Replacing the featured image removes the featured image from the post
  B. Replacing the featured image leaves the featured image in place in Post Settings (this is the current behavior)
  C. Replacing the featured image updates the post's featured image
Intuitively, I thought option C would be how this would work. Maybe I'd associated the "Featured" banner too closely to the Image block itself, when it's in fact an attribute of the _image_. Again, not a blocker, just wanted to flag this.

---

@SiobhyB, great work here and I appreciate the easy to ready testing instructions 👍. I haven't had a chance to look at the code but will do that next.

guarani commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -430,6 +433,54 @@ export class ImageEdit extends Component {
+		setFeaturedImage( mediaId );
+	}
+
+	getSetFeaturedButton( isFeaturedImage ) {
```
The get-set in the name `getSetFeaturedButton` is a bit confusing to me. Could it be replaced with `getFeaturedButtonPanel` or similar?

SiobhyB replied:
Absolutely, I've gone ahead to update the name in https://github.com/WordPress/gutenberg/pull/31415/commits/d17376307ca4f15cc694db95933c35dbb8e3181b. :D

SiobhyB committed:
Update function name for clarity

Following feedback, the getSetFeaturedButton function has been re-named to getFeaturedButtonPanel for clarity.

SiobhyB commented:
Thanks so much for your thorough review, Paul! I'll go ahead to work through it all and address feedback from each "test case" in a separate comment.

SiobhyB commented:
## Response to Feedback from Test Case 1

> The "Set as Feature Image" option is off-screen for me on iPhone 11 Pro Max simulator when using an image that had a link set, so its discoverability seems low in certain situations (which reminded me of [this post](https://tyler.io/perfectly-cropped/) 😄).

Ah, thank you for sharing that post! I enjoyed the writer's perspective and it reminded me of similar encounters with users, who weren't aware of certain features due to a lack of scroll indicators. As the issue can be found on all BottomSheets with overflowing content, it seems like this may be larger than this specific PR. I've gone ahead to create an enhancement request at https://github.com/WordPress/gutenberg/issues/32409. 

Let me know if you think any more action is required around this specific PR.

> The small edit/replace button that was previously shown inside an image canvas is now shown outside. It looks like this bug isn't present on trunk.

I wasn't able to reproduce an issue specific to this branch. Did you test using the same image, with the same size set? 

I do see that the edit/replace button appears outside of the image in cases where the size is set to anything other than `Full Size` (via the block's settings), but this was true in trunk in my testing also. 

> When the message "Set/Removed as Feature Image" is shown for a few seconds, just before it's removed, it seems to flash momentarily. I'm not sure if this is a symptom of something being called twice for example, so I just want to flag it for you (at 0:05s in the video below)

I'm able to replicate the flickering but having a difficult time tracking down why it's happening. I'm only able to replicate it when a featured image is set/removed straight away (i.e. the replace dialog isn't used), so I wonder if it's something related to the BottomSheet dismissal. 🤔 I'll keep exploring.

SiobhyB commented:
## Response to Feedback from Test Case 2

> The bottom (default) button in the action sheet that's shown when replacing a featured image should be "Cancel" At the moment, both options shown are "Replace"

This was fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/16426/commits/02db10415a3324ed44018c23f55dd32335bcd415, thank you so much for spotting that!

> Looking at the screenshot above, I'm not sure if the action sheet's title should be a question (i.e. "Replace?"). Looking at other parts of the app, I notice the edit/replace button's action sheet doesn't use a question in its title. An alternative could be "Featured image already set" (or "Featured image options"). I'm fine leaving it as-is, just wanted to share the thought.

I appreciate the extra context around how other titles work in the app. I like `Featured Image Already Set` (using title case), I've gone ahead to use that wording in and updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/16426/commits/8063237c26f010b666da712e98add89761a94474. I've also updated the Android side in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/1eb55fa123df127a85a302f0e8c75f5b94eed1d4.

> When a post already has a featured image set via Post Settings and the user goes to set a different image via the Image block and is presented with the replace action sheet, I think some users might be confused because they don't see any other image in the post marked as Featured. 

I agree that there may be some confusion around _which_ image is set as featured. Thomas mentioned this in a design review, also, and I've opened an enhancement request at https://github.com/WordPress/gutenberg/issues/30410. Although I'll be focusing on other areas of code after this project, I've added it to a list that I hope to get to _someday_, perhaps a future HACK week when I've built up my knowledge a bit more. :) 

> "Feature Image" in the "Set/Removed as Feature Image" messages maybe should be lowercase i.e. "Set/Removed as feature image" (it's up to you, I'm open to what you think is best here).

I initially viewed these as headings of sorts, which is why I went with title case, but I think that's a bit of a debatable classification and you're right that title case isn't really appropriate, especially as other notices don't use title case in the app. I've gone ahead to update this for iOS in https://github.com/wordpress-mobile/WordPress-iOS/pull/16426/commits/7e7a291001f0d06b4915c7205aeb718ac6b34b03 and for Android in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/c1f68422ddf3e4b0b3071b37eb79067bad203201.

SiobhyB commented:
## Response to Feedback from Test Case 5

```
Here things worked as described in the test case, but my expectation was that replacing a featured image would update the post's featured image to the new image. In this scenario I can think of three options:
A. Replacing the featured image removes the featured image from the post
B. Replacing the featured image leaves the featured image in place in Post Settings (this is the current behavior)
C. Replacing the featured image updates the post's featured image
Intuitively, I thought option C would be how this would work. Maybe I'd associated the "Featured" banner too closely to the Image block itself, when it's in fact an attribute of the image. Again, not a blocker, just wanted to flag this.
```

This is really useful feedback that matches some other internal chatter, thank you! I'll keep monitoring feedback like this and consider tracking it via a new GitHub issue if it continues to be a theme. 🙇‍♀️

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

guarani commented:
> ## Response to Feedback from Test Case 1

> Ah, thank you for sharing that post! I enjoyed the writer's perspective and it reminded me of similar encounters with users, who weren't aware of certain features due to a lack of scroll indicators. As the issue can be found on all BottomSheets with overflowing content, it seems like this may be larger than this specific PR. I've gone ahead to create an enhancement request at #32409.
> 
> Let me know if you think any more action is required around this specific PR.

Thanks for making that issue! I agree it isn't specific to this PR so happy to see it left out for now.

> I wasn't able to reproduce an issue specific to this branch. Did you test using the same image, with the same size set?
> 
> I do see that the edit/replace button appears outside of the image in cases where the size is set to anything other than `Full Size` (via the block's settings), but this was true in trunk in my testing also.

I'm [having trouble](https://github.com/wordpress-mobile/WordPress-iOS/pull/16426#issuecomment-853429025) with the iOS build at the moment, but once that's done I'll come back and test this to make sure I see the same thing on this branch and on `trunk`.

> I'm able to replicate the flickering but having a difficult time tracking down why it's happening. I'm only able to replicate it when a featured image is set/removed straight away (i.e. the replace dialog isn't used), so I wonder if it's something related to the BottomSheet dismissal. 🤔 I'll keep exploring.

I've seen some minor quirks with that notice. For example, if you delete a couple of blocks one-after-the-other, the "Block removed" notices get queued up which is probably unexpected (at least in that scenario). I haven't seen this particular flickering issue anywhere else though.

> ## Response to Feedback from Test Case 2

> This was fixed in [wordpress-mobile/WordPress-iOS@02db104](https://github.com/wordpress-mobile/WordPress-iOS/commit/02db10415a3324ed44018c23f55dd32335bcd415), thank you so much for spotting that!

No worries!

> I appreciate the extra context around how other titles work in the app. I like `Featured Image Already Set` (using title case), I've gone ahead to use that wording in and updated in [wordpress-mobile/WordPress-iOS@8063237](https://github.com/wordpress-mobile/WordPress-iOS/commit/8063237c26f010b666da712e98add89761a94474). I've also updated the Android side in [wordpress-mobile/WordPress-Android@1eb55fa](https://github.com/wordpress-mobile/WordPress-Android/commit/1eb55fa123df127a85a302f0e8c75f5b94eed1d4).

Sounds good!

> I agree that there may be some confusion around _which_ image is set as featured. Thomas mentioned this in a design review, also, and I've opened an enhancement request at #30410. Although I'll be focusing on other areas of code after this project, I've added it to a list that I hope to get to _someday_, perhaps a future HACK week when I've built up my knowledge a bit more. :)

Looks great, thanks for creating that issue.

> I initially viewed these as headings of sorts, which is why I went with title case, but I think that's a bit of a debatable classification and you're right that title case isn't really appropriate, especially as other notices don't use title case in the app. I've gone ahead to update this for iOS in [wordpress-mobile/WordPress-iOS@7e7a291](https://github.com/wordpress-mobile/WordPress-iOS/commit/7e7a291001f0d06b4915c7205aeb718ac6b34b03) and for Android in [wordpress-mobile/WordPress-Android@c1f6842](https://github.com/wordpress-mobile/WordPress-Android/commit/c1f68422ddf3e4b0b3071b37eb79067bad203201).

🙇 

> ## Response to Feedback from Test Case 5

> This is really useful feedback that matches some other internal chatter, thank you! I'll keep monitoring feedback like this and consider tracking it via a new GitHub issue if it continues to be a theme. 🙇‍♀️

Glad to know I'm not alone raising that feedback. 
Thanks again @SiobhyB.

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

SiobhyB committed:
Remove androidOnly const

With this PR, the featured button will work on iOS and this const will no longer be needed.

SiobhyB commented:
> I've seen some minor quirks with that notice. For example, if you delete a couple of blocks one-after-the-other, the "Block removed" notices get queued up which is probably unexpected (at least in that scenario). I haven't seen this particular flickering issue anywhere else though.

Thanks for this extra context! I did some testing and am also able to reproduce some quirkiness with the notice on Android. If I'm not able to find a straightforward fix as part of this PR, I might lean towards creating a separate issue. But I'll spend a bit more time experimenting tomorrow first. 🙇‍♀️

guarani commented:
> I'm [having trouble](https://github.com/wordpress-mobile/WordPress-iOS/pull/16426#issuecomment-853429025) with the iOS build at the moment, but once that's done I'll come back and test this to make sure I see the same thing on this branch and on `trunk`.

I think your comment earlier was spot on, it has to do with the Image size setting and has nothing to do with this PR.

guarani commented:
I gave this another test today and spotted one issue: Sometimes the Featured Image section in Post Settings displays an "Upload Failed. Tap for options." message when I select and then deselect featured images. Then when I retry from the Post Settings screen, it loads an image from a previous post (not the image from the post I'm currently on):

https://user-images.githubusercontent.com/1898325/120732544-82f1b200-c4b3-11eb-8d89-a92807a45914.mov

SiobhyB committed:
Add upload check to 'canImageBeFeatured'

An image can't be set as featured unless it's been successfully uploaded to the site's media library, but the 'canImageBeFeatured' check didn't account for this.

A temporary negative ID is given to images that are in progress or fail to upload. With this commit, a check for whether an image's ID is more than zero is added to 'canImageBeFeatured' in order to account for those times.

Note, 'this.state.isUploadInProgress' isn't used, as it doesn't account for times where an upload fails. Instead, a ID check acts as a catch all.

SiobhyB commented:
@guarani, thank you for spotting that! I _think_ what might be happening is this:

* An image is given a temporary negative ID while it's in the process of uploading e.g. `-2838582747`
* As the featured button is available while an image is uploading, it's technically possible to send that negative ID back to the JS side. 
* If the negative ID is sent back to the JS side then it'd cause some issues, including the wrong image displayed in Post Settings.

I've added a check to ensure the button only displays when an image's ID is greater than zero in https://github.com/WordPress/gutenberg/pull/31415/commits/60d205c4ebbd4f137460448f136076dd5725876a as a way to avoid this.

That said, I have some uncertainty as it _seems_ like the image in your video is fully uploaded and I had a bit of trouble reliably replicating. I _was_ able to replicate, but only following a lot of trial and error with forcing an upload to fail while in progress (by turning the WiFi off). 

I'm not able to replicate the issue any more but, if you're still able to replicate on your side even with the latest change, would you be able to add the following logging within the `render()` function [in the image block's `edit.native.js` file](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L484) for me?

```
console.log( "Image's ID: " + attributes.id );
console.log( "Featured image's ID: " + featuredImageId );
```

I'd be interested to know what the output is after you first click the `Set as Featured` button and then again after clicking the `Remove as Featured` button. (Ideally, both values should be the same after you first click the button, with `featuredImageId` changing to zero after you remove it.)

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

SiobhyB committed:
Merge branch 'add/featured-functionality-to-image-block-ios' into add/set-as-featured-button-ios

guarani reviewed (approved):
Testing https://github.com/wordpress-mobile/WordPress-iOS/pull/16426/commits/8f473649cfb790a736257fb742e8bfea7c31adc5 on iPad 8th gen:

- [x] Test Case 1: General Flow
- [x] Test Case 2: "Replace Existing" Flow
On iPad, cancel buttons are not rendered (which is expected I think) so tapping outside the action sheet works instead of cancelling.
- [x] Test Case 3: "Editor Loads" Flow
- [x] Test Case 4: "Post Settings" Flow
With regard to setting a featured image via Post Settings and _then_ adding the image to the editor via the image block, the featured image banner only shows up if I add the image from the WP Media Library, not the device. I think this is expected though, since adding an image from the device is treated as a different image to the one that's in the WP Media Library, even if they're the same image.
- [x] Test Case 5: "Replace in Image Block" Flow
Same applies here where replacing an image only restores the featured image banner if the image is re-added via the WP Media Library.

I don't think any of the above are issues, just wanted to note them here. 
Looks great and works great, thanks @SiobhyB!
